From: Carlos Garnacho <carlosg@gnome.org>
Date: Tue, 3 Dec 2024 20:41:44 +0100
Subject: wayland: Emit wl_touch.cancel in seat focus handling

We try to emit wl_touch.cancel in the situations where grabbing input
would take input away from the wl_touch interface. This however does
not play fine with grabs induced by wayland interfaces (e.g. xdg popups).

A more natural place to handle this is the MetaWaylandSeats' default
input handler, specifically the focus() vmethod. Here, we may know if the
focus surface matches the current surface as picked by MetaWaylandTouch,
and if it does not, either a grab or another input handler stole input
away from the default event handler.

A case where this will notably not happen anymore is in the transition
to a xdg_popup grab, since the grabbing client will still be handling
input, and touch input will not be unfocused away despite the transition
to a grab.

Closes: https://gitlab.gnome.org/GNOME/mutter/-/issues/3752
Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4166>

Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/3752
Bug-Ubuntu: https://bugs.launchpad.net/bugs/2091159
---
 src/wayland/meta-wayland-input.c | 4 ----
 src/wayland/meta-wayland-seat.c  | 6 +++++-
 2 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/src/wayland/meta-wayland-input.c b/src/wayland/meta-wayland-input.c
index 066aa9b..69296de 100644
--- a/src/wayland/meta-wayland-input.c
+++ b/src/wayland/meta-wayland-input.c
@@ -187,10 +187,6 @@ meta_wayland_input_invalidate_all_focus (MetaWaylandInput    *input,
         meta_wayland_event_handler_invalidate_focus (handler, device, l->data);
     }
 
-  if (meta_wayland_seat_has_touch (seat) &&
-      (flags & INVALIDATE_FOCUS_FLAG_CANCEL_TOUCH) != 0)
-    meta_wayland_touch_cancel (seat->touch);
-
   g_hash_table_iter_init (&iter, seat->tablet_seat->tablets);
   while (g_hash_table_iter_next (&iter, (gpointer*) &device, NULL))
     {
diff --git a/src/wayland/meta-wayland-seat.c b/src/wayland/meta-wayland-seat.c
index 8e592d6..296702f 100644
--- a/src/wayland/meta-wayland-seat.c
+++ b/src/wayland/meta-wayland-seat.c
@@ -216,7 +216,11 @@ default_focus (MetaWaylandEventHandler *handler,
   ClutterInputCapabilities caps;
 
   if (sequence)
-    return;
+    {
+      if (surface != meta_wayland_touch_get_surface (seat->touch, sequence))
+        meta_wayland_touch_cancel (seat->touch);
+      return;
+    }
 
   caps = clutter_input_device_get_capabilities (device);
 
